doc/neps/nep-0023-backwards-compatibility.rst

```diff
@@ -19,46 +19,206 @@ processes for individual cases where breaking backwards compatibility
+
+Out of scope are:
+
+- Making concrete decisions about deprecations of particular functionality.
+- NumPy's versioning scheme.
+
+
+General principles
+------------------
+
+When considering proposed changes that are backwards incompatible, the
+main principles the NumPy developers use when making a decision are:
+
+1. Changes need to benefit users more than they harm them.
+2. NumPy is widely used so breaking changes should by default be assumed to be
+   fairly harmful.
+3. Decisions should be based on data and actual effects on users and downstream
+   packages rather than, e.g., appealing to the docs or for stylistic reasons.
+4. Silently getting a wrong answer is much worse than getting a loud error.
+
+When assessing the costs of proposed changes, keep in mind that most users do
+not read the mailing list, do not look at deprecation warnings, and sometimes
+wait more than one or two years before upgrading from their old version. And
+that NumPy has millions of users, so "no one will do or use this" is very
+likely incorrect.
+
+Benefits include improved functionality, usability and performance, as well as
+lower maintenance cost and improved future extensibility.
+
+Fixes for clear bugs are exempt from this backwards compatibility policy.
+However in case of serious impact on users (e.g. a downstream library doesn't
+build anymore or would start giving incorrect results), even bug fixes may have
+to be delayed for one or more releases.
+
+
+Strategies related to deprecations
+----------------------------------
+
+Getting hard data on the impact of a deprecation of often difficult. Strategies
+that can be used to assess such impact include:
+
+- Use a code search engine ([1]_) or static ([2]_) or dynamic ([3]_) code
+  analysis tools to determine where and how the functionality is used.
+- Testing prominent downstream libraries against a development build of NumPy
+  containing the proposed change to get real-world data on its impact.
+- Making a change in master and reverting it, if needed, before a release. We
+  do encourage other packages to test against NumPy's master branch, so this
+  often turns up issues quickly.
+
+If the impact is unclear or significant, it is often good to consider
+alternatives to deprecations. For example, discouraging use in documentation
+only, or moving the documentation for the functionality to a less prominent
+place or even removing it completely. Commenting on open issues related to it
+that they are low-prio or labeling them as "wontfix" will also be a signal to
+users, and reduce the maintenance effort needing to be spent.
+
+
+Implementing deprecations and removals
+--------------------------------------
+
+Deprecation warnings are necessary in all cases where functionality
+will eventually be removed.  If there is no intent to remove functionality,
+then it should not be deprecated either. A "please don't use this" in the
+documentation or other type of warning should be used instead.
```

Comment:
When it has been decided that functionality should be removed this functionality this is a "Deprecation" and goes through the following deprecation process...

If you decide to introduce an "alternatives" section, you could move the "please don't use this" part there.

(Just thought it may be nice to try and connect it a bit with the previous section.)

Reply:
It does seem a bit odd to leave functionality in place that won't be used.  This increases maintenance burden for no good reason.  I think this is the case for some of our "we would love to remove this but don't think we can get away with it" functions.

Reply:
I'll change to "please don't use this for new code". We have a lot of functionality like that. I don't think it's odd, more powerful or intuitive APIs can get added, that shouldn't mean you just rip out the old stuff.

Reply:
Also adding "and the documentation can be organized such that the preferred alternative is more prominently shown.". That should cover it I think.